### DISCUSSION TITLE
Building Presently for iOS

### DISCUSSION THREAD

**Discussion by alisonthemonster** [alisonthemonster/GratitudeJournal#165] [General]
Users ask me all the time if there is an iOS version of Presently. At the moment there is not, but it doesn't have to be this way!

Presently is 100% open source and many features and contributions have come from users here on Github, if you are interested in building Presently for iOS please make yourself known here! 

We would also love some iOS design love here to change some very Android-y components over to Human Interface Guidelines!

**Some guidelines for the iOS app**
- Fully open source
- Written in Swift and following modern architecture patterns (Clean, unidirectional, strong separation of concerns, etc)
- Tested!! Unit tests and ideally some sort of screenshot testing so the UI can be validated by a non-iOS dev would be fantastic
- Entries stored in app level local storage
- Crash reporting and analytics for monitoring app health and usage
- CI to ensure app builds on every commit and that tests pass and coverage is met
- Eventually full feature parity

**Features listed in priority order:**
- [ ] Time line with entries listed and hints to guide users to write for today and yesterday
- [ ] Entry screen to enter their gratitude
- [ ] Daily reminder notification
- [ ] Contact link to email us
- [ ] Privacy policy link to webview
- [ ] Terms + conditions link to webview
- [ ] OSS libraries list
- [ ] Calendar to jump days in the past
- [ ] Milestones in the timeline and dialogs to appear when milestones are acheived
- [ ] Export/Import to CSV
- [ ] Link to FAQ webview
- [ ] THEMES! 
- [ ] Dropbox automatic backups in the background
- [ ] Biometric or PIN locking
- [ ] Search
- [ ] Small customizations to timeline + calendar (adding day of week, changing first day of week in calendar)

If this discussion gets enough interest I can go through and create a Kanban board and some more robust contribution rules and work out some timelines and plans!

Cheers to open source 🥂

**Comment by alisonthemonster** [comment]
Funny story, someone once totally ripped off Presently and wrote an iOS version. Had to contact Apple and get it removed! 

It had lots of typos and the design was slightly off. Luckily a few friends noticed this and contacted me about it and we were able to get it removed quite quickly. 
You can see in this screenshot they even copied the screenshots and I had for this repo! `Sesame seeds and sushi rice` 😆  They also copied the store listing exactly. 
![image](https://user-images.githubusercontent.com/10744793/104775851-e742ed80-572d-11eb-86fc-c232f065d5b8.png)
<img width="771" alt="Screen Shot 2021-01-15 at 12 36 08 PM" src="https://user-images.githubusercontent.com/10744793/104776059-4acd1b00-572e-11eb-8d65-68c6228d4e14.png">

**Comment by polbins** [comment]
Hey! It's good to see that this is being considered!
I actually am using an iPhone as a daily driver nowadays so I'm not able to use Presently anymore.

I hope someone will take a lead on this one and I can help with integrating some of the features. :)

**Comment by peakvalleytech** [comment]
Hi Alison, glad to hear that you were able to deal justice. 

I actually know some iOS development. I would be happy to get the ball rolling on this. How far have you gotten so far?

**Reply by alisonthemonster** [reply]
Hi there! 
This is as far as I have gotten! Would love to get a github project started and create some issues there for tracking if we get enough traction!

**Reply by peakvalleytech** [reply]
Ok, i'll see if I can get a prototype up and running. Let me know when you set up the project, so I can send a PR.

**Comment by davidwilliston** [comment]
Why not use flutter?

**Comment by devandapaige** [comment]
I know it's been a while, but I'd be interested in helping in any way I can. This is honestly one of my favorite apps, and I'd love to help bring it to the iOS world. I'm still a baby dev, but shoot me a message if I can help!

**Comment by Eloren1** [comment]
While the developer removes clones from the App Store and does not offer alternatives himself, I decided to use the app called "Day One". They have import function, but the formats are slightly different. So I made a python script for converting the export file from Presently to another format. When importing to Day One, it will appear as a separate list on the left, you can delete the rest.

```
try:
    import pandas as pd
except ImportError:
    import subprocess
    subprocess.check_call(["pip", "install", "pandas"])
    import pandas as pd
import uuid
import os

# SETTINGS (SET YOUR VALUES)
gmt_offset  = 3  # your current GMT shift for correct date/time
df_old = pd.read_csv('OLD.csv')  # your file in the same folder

def convert_row(row):
    utc_hour = (24 - gmt_offset) % 24
    timestamp = f"{(pd.to_datetime(row['entryDate']) - pd.Timedelta(days=1)).strftime('%Y-%m-%d')}T{utc_hour:02}:00:00.000Z"
    return {
        'date': timestamp,
        'text': row['entryContent'],
        'uuid': uuid.uuid4().hex.upper(),
        'duration': 0,
        'starred': False,
        'isPinned': False,
        'editingTime': 0.0,
        'modifiedDate': timestamp,
        'creationDevice': '',
        'timeZoneIdentifier': '','authorName': '','conditionsDescription': '','moonPhase': '','moonPhaseCode': '','pressureMB': '','relativeHumidity': '','sunriseDate': '','sunsetDate': '','temperatureCelsius': '','visibilityKM': '','weatherCode': '','weatherServiceName': '','windBearing': '','windChillCelsius': '','windSpeedKPH': '','address': '','administrativeArea': '','altitude': '','country': '','floorLevel': '','foursquareID': '','heading': '','latitude': '','localityName': '','longitude': '','placeName': '','speed': '','locationTimeZoneName': '','locationUserLabel': '','locationUserType': '','regionCenterLatitude': '','regionCenterLongitude': '','regionRadius': '','activityName': '','ignoreStepCount': '','stepCount': '','mediaMD5s': '',
    }

converted = pd.DataFrame([convert_row(row) for _, row in df_old.iterrows()])
converted.to_csv('NEW_converted.csv', index=False)

print("Saved as NEW_converted.csv")
os.system("pause")
```

**Comment by mysocalledmike** [comment]
Hey Alison!

Long-time Presently user here. My Android phone just died and I switched to iPhone, which means I've lost the app I've used every day for the past 8 years (literally) to journal. Rather than migrate to Day One, I'd like to do what this discussion originally proposed: build a native iOS port. I know from your comment that you've been burned by a clone before, so just thoughts on how this could work:

**On credit and control.** This would be your project, ported. You get final say on whether it ships and where it lives (your GitHub org, mine as a mirror, whatever you want). I'd credit everything as an "iOS port of Presently by @alisonthemonster." and themes get credited to the original designers. I also wouldn't publish to the App Store without your blessing, and if you'd prefer to be the publisher with me as developer, we can do that too.

**On licensing.** I noticed the repo doesn't currently have a LICENSE file. Since you've publicly committed to "always be open source," would you be open to adding one to the Android repo (MIT, Apache-2.0, and GPL-3.0 are the common choices) so the iOS port can inherit the same license cleanly? Happy to open a PR with whichever you prefer.

**On approach.** 100% native Swift/SwiftUI, iOS 17+. Following your 2021 guidelines: modern architecture, tests, CI, local-first storage, crash reporting. Aiming for eventual full feature parity, phased in roughly your original priority order (timeline + entry + reminder first, themes and biometric lock next, search and backup later). One potential divergence is to replace Dropbox with iCloud sync since it's more seamless, but open to your take.

No pressure to respond quickly. If you want to talk more in detail, feel free to write back here (or on LinkedIn - we have a thread there too). 

Thank you again for building Presently!
–Mike
